include/vinecopulib/vinecop/implementation/class.ipp

```diff
@@ -780,8 +782,8 @@ Vinecop::pdf(Eigen::MatrixXd u, const size_t num_threads) const
         tools_interface::check_user_interrupt(edge % 100 == 0);
         // extract evaluation point from hfunction matrices (have been
         // computed in previous tree level)
-        Bicop edge_copula = get_pair_copula(tree, edge);
-        auto var_types = edge_copula.get_var_types();
+        Bicop* edge_copula = &pair_copulas_[tree][edge];
```

Comment:
I like the idea of not copying, but maybe a smart pointer rather than a raw pointer?

Reply:
good idea

Reply:
Actually it is not. The 'benefit' of a smart pointer would be that the pointed-to object is destroyed when the pointer goes out of scope, but that's not what we want here. The pointed-to object is destroyed whenever `pair_copulas_` is destroyed.